content/blog/new-graphic-features/index.Rmd

```diff
@@ -0,0 +1,301 @@
+
+```{r}
+circle <- circleGrob(
+  x = c(0.2, 0.8), 
+  r = 0.3
+)
+rect <- rectGrob(
+  width = unit(0.7, 'snpc'), 
+  height = unit(0.7, 'snpc'), 
+  vp = viewport(angle = 45)
+)
+mask <- gTree(children = gList(circle, rect), gp = gpar(fill = "#00000066", col = NA))
+grid.rect(
+  x = c(0.25, 0.25, 0.75, 0.75), 
+  y = c(0.25, 0.75, 0.75, 0.25), 
+  width = 0.5,
+  height = 0.5,
+  gp = gpar(fill = c("steelblue", "firebrick")),
+  vp = viewport(mask = mask)
+)
+```
+
+As we see above, the areas in the mask where nothing is drawn have an opacity of 0, meaning that whatever is being drawn by the rectangle grob in these areas will be invisible. We also see that opacity is compounded by overlaying shapes as the areas covered both by the circle and the square in the mask has a higher opacity.
+
+## Future features
+
+It is of course dangerous to promise what the future brings. However, I do know of a few features adjacent to what have been discussed above that might make sense to know about.
+
+When it comes to clipping paths there is currently a lack of way to describe how multiple shapes are combined, since the fill rule is implicitly "winding" and you have no control over the direction the graphic device trace circles and rectangles. Work is already being done to let you control this from grid so it will become easier to e.g. punch out holes in a clipping path by overlaying two grobs.
+
+As noted in the discussion about masks, only alpha masks are currently possible. However, producing an exact transparency through compounded shapes can be tough because of the way opacity combines. In the future there will also be support for luminosity masks and this should greatly improve the user experience of this feature in my opinion.
+
+Still, the main takeaway from all of the above is that the graphic engine is once again a living breathing code-base with big user-facing features on the horizon.
+
+## The ggplot2 implications
+
+While ggplot2 is using grid underneath it's grammar API, these features are generally not directly available in ggplot2. This is because most of these features are not directly applicable to the current API. Both gradient and patterns are obvious candidates for extensions of the ggplot2 API, but for now the grid API doesn't support a vector of patterns/gradients. Once this limitation is removed (it is in the works), we will need to figure out how scaling of these more flexible fill types should work. The starting point is of course to allow mapping from one data-value to a predefined pattern/gradient, but it would be interesting to think about how to map data-values to features of the pattern/gradient, e.g. have the gradient defined by two or more columns that all maps to different colors. Some of this work and exploration is already happening in [ggpattern](https://coolbutuseless.github.io/package/ggpattern), which could form the basis of future ggplot2 support.
```

Comment:
```suggestion
While ggplot2 uses grid underneath it's grammar API, these features are generally not directly available in ggplot2. This is because most of these features are not directly applicable to the current API. Both gradients and patterns are obvious candidates for extensions of the ggplot2 API. But, for now, the grid API doesn't support a vector of patterns/gradients. Once this limitation is removed (it is in the works), we will need to figure out how scaling of these more flexible fill types should work. The starting point is, of course, to allow mapping from one data-value to a predefined pattern/gradient, but it would be interesting to think about how to map data-values to features of the pattern/gradient, e.g. have the gradient defined by two or more columns that all maps to different colors. Some of this work and exploration is already happening in [ggpattern](https://coolbutuseless.github.io/package/ggpattern), which could form the basis of future ggplot2 support.
```